File: mac-app/QuartzTeachers/Models/PreviewLeases.swift
import Foundation

/// Hands out preview ports, one per running preview, across every window.
///
/// The container publishes ports 8081–8084, so up to four previews can run
/// at once — a teacher comparing sections side by side. Each preview leases
/// a port when it starts and returns it when it stops; the same section can
/// only be previewed in one place, because two builds of one section would
/// race over the same output folder.
@MainActor
enum PreviewLeases {

    // MARK: - Types

    struct Lease: Equatable {
        let port: Int
        let folderPath: String
        let courseCode: String
        let sectionNumber: Int
    }

    /// Why a preview could not start, in words a teacher can act on.
    enum Problem: LocalizedError {
        case sectionAlreadyPreviewed(String, Int)
        case allPortsBusy

        var errorDescription: String? {
            switch self {
            case .sectionAlreadyPreviewed(let code, let section):
                return "Section \(section) of \(code) is already being previewed in another window. Stop that preview first, or work with it there."
            case .allPortsBusy:
                return "Four previews are already running, which is the most that can run at once. Stop one, then try again."
            }
        }
    }

    // MARK: - Stored properties

    /// The ports the container publishes.
    static let availablePorts: [Int] = [8081, 8082, 8083, 8084]

    /// The previews currently running, across all windows.
    private(set) static var active: [Lease] = []

    // MARK: - Functions

    /// Leases a port for a preview of one section, refusing politely when
    /// the section is already live elsewhere or every port is taken.
    static func lease(folderPath: String, courseCode: String, sectionNumber: Int) throws -> Lease {
        for existing in active {
            let samePlace: Bool = existing.folderPath == folderPath
                && existing.courseCode == courseCode
                && existing.sectionNumber == sectionNumber
            if samePlace {
                throw Problem.sectionAlreadyPreviewed(courseCode, sectionNumber)
            }
        }

        var takenPorts: [Int] = []
        for existing in active {
            takenPorts.append(existing.port)
        }
        for port in availablePorts {
            if !takenPorts.contains(port) {
                let lease: Lease = Lease(
                    port: port,
                    folderPath: folderPath,
                    courseCode: courseCode,
                    sectionNumber: sectionNumber
                )
                active.append(lease)
                return lease
            }
        }
        throw Problem.allPortsBusy
    }

    /// Returns a lease when its preview stops, whatever the reason.
    static func release(_ lease: Lease) {
        var remaining: [Lease] = []
        for existing in active {
            if existing != lease {
                remaining.append(existing)
            }
        }
        active = remaining
    }

    /// Starts from nothing — for tests.
    static func reset() {
        active = []
    }
}

File: windows-app/Plantoir.Core/Models/PreviewLeases.cs
namespace Plantoir.Core.Models;

/// <summary>
/// Hands out preview ports across every window. The four ports are the
/// CONTAINER-side ports; the launcher maps them into a probed host block,
/// so the app never assumes the host port — it parses the announced address.
/// </summary>
public static class PreviewLeases
{
    public static readonly IReadOnlyList<int> AvailablePorts = new[] { 8081, 8082, 8083, 8084 };

    public sealed record Lease(int Port, string FolderPath, string CourseCode, int SectionNumber);

    public sealed class LeaseRefusedException(string message) : Exception(message);

    private static readonly List<Lease> _active = new();
    private static readonly object _gate = new();

    public static IReadOnlyList<Lease> Active { get { lock (_gate) return _active.ToList(); } }

    public static Lease Take(string folderPath, string courseCode, int sectionNumber)
    {
        lock (_gate)
        {
            // Two builds of one section would race over the same output
            // folder — but the same section in a DIFFERENT folder is the
            // compare-two-years case, and is allowed.
            if (_active.Any(l => l.FolderPath == folderPath &&
                                 l.CourseCode == courseCode &&
                                 l.SectionNumber == sectionNumber))
                throw new LeaseRefusedException(
                    $"Section {sectionNumber} of {courseCode} is already being previewed in another window. Stop that preview first, or work with it there.");

            // Ports are per-container, one container per folder — only
            // same-folder previews contend.
            var taken = _active.Where(l => l.FolderPath == folderPath).Select(l => l.Port).ToHashSet();
            foreach (int port in AvailablePorts)
            {
                if (taken.Contains(port)) continue;
                var lease = new Lease(port, folderPath, courseCode, sectionNumber);
                _active.Add(lease);
                return lease;
            }
            throw new LeaseRefusedException(
                "Four previews of this folder are already running, which is the most that can run at once. Stop one, then try again.");
        }
    }

    public static void Release(Lease lease)
    {
        lock (_gate) _active.Remove(lease);
    }

    public static void Release(string folderPath, string courseCode, int sectionNumber)
    {
        lock (_gate)
        {
            _active.RemoveAll(l => l.FolderPath == folderPath &&
                                  string.Equals(l.CourseCode, courseCode, StringComparison.OrdinalIgnoreCase) &&
                                  l.SectionNumber == sectionNumber);
        }
    }

    public static void Reset()
    {
        lock (_gate) _active.Clear();
    }
}
